src/bottom_up_iterators/bottom_up_depth_iterator.jl

```diff
@@ -0,0 +1,200 @@
+* `new_programs::Vector{RuleNode}`: Programs generated at the current depth level. Will be appended to the `bank` when starting a new depth level.
+* `rules::Queue{Int}`: Grammar rules left to be used at the current depth level. 
+* `depth::Int`: Current depth.
+"""
+mutable struct BUDepthData <: BottomUpData
+    nested_rulenode_iterator::NestedRulenodeIterator
+    new_programs::Vector{RuleNode}
+    rules::Queue{Int}
+    depth::Int
+end
+
+"""
+	order(iter::BUDepthIterator)::Queue{Int}
+
+Returns the non-terminal rules in the order in which they appear in the grammar.
+"""
+function order(
+    iter::BUDepthIterator
+)::Queue{Int}
+    grammar::ContextSensitiveGrammar = get_grammar(iter.solver)
+    rules::Queue{Int} = Queue{Int}()
+
+    for (rule_index, is_terminal) ∈ enumerate(grammar.isterminal)
+        if !is_terminal
+            enqueue!(rules, rule_index)
+        end
+    end
+
+    return rules
+end
+
+"""
+	init_bank(iter::BUDepthIterator)::BUDepthBank
+
+Returns an initialized object of type `BUDepthBank`. For each symbol in the grammar (i.e. key in the dictionary), an empty `Vector{RuleNode}` (i.e. value in the dictionary) is allocated.
+"""
+function init_bank(
+    iter::BUDepthIterator
+)::BUDepthBank
+    grammar::ContextSensitiveGrammar = get_grammar(iter.solver)
+    rulenodes_by_symbol::Dict{Symbol, Vector{RuleNode}} = Dict{Symbol, Vector{RuleNode}}()
+
+    for symbol ∈ grammar.types
+        rulenodes_by_symbol[symbol] = Vector{RuleNode}()
+    end
+
+    return BUDepthBank(rulenodes_by_symbol)
+end
+
+"""
+    init_data(iter::BUDepthIterator)::BUDepthData
+
+Returns an initialized object of type `BUDepthData`. The initialization consists of the following:
+* `nested_rulenode_iterator` is set to an empty `NestedRulenodeIterator`.
+* `rules` contains the indeces of terminal rules in the grammar.
+* `new_programs` is an empty `Vector{RuleNode}`.
+* `depth` is set to 1.
+"""
+function init_data(
+    iter::BUDepthIterator
+)::BUDepthData
+    grammar::ContextSensitiveGrammar = get_grammar(iter.solver)
+
+    empty_nested_iterator::NestedRulenodeIterator = NestedRulenodeIterator()
+    depth::Int = 1
+
+    rules::Queue{Int} = Queue{Int}()
+    for (rule_index, is_terminal) ∈ enumerate(grammar.isterminal)
+        if is_terminal
+            enqueue!(rules, rule_index)
+        end
+    end
+
+    return BUDepthData(empty_nested_iterator, Vector{RuleNode}(), rules, depth)
+end
+
+"""
+    create_program!(iter::BUDepthIterator, bank::BUDepthBank, data::BUDepthData)::Union{Nothing, RuleNode}
+
+Returns the next program in `BUDepthIterator`'s iteration. Performs the following steps:
+1. Check whether `data.nested_rulenode_iterator` contains any program.
+2. Otherwise, pick the next rule from `data.rules` and generate all combinations of `RuleNodes` from the bank that could be combined using this rule.
+3. If `data.rules` is empty, call `_increase_depth!` and go to step 2.
```

Comment:
I would prefer the docstring to not contain the steps of the function like this with all of the implementation details—that's what the source is for. Maybe we move the comments into inline comments and keep this part more general. I just worry the two will get out of sync of there are changes.